src/shogun/evaluation/CrossValidation.cpp

```diff
@@ -264,32 +264,57 @@ float64_t CCrossValidation::evaluate_one_run()
 		m_machine->set_store_model_features(true);
 
 		/* do actual cross-validation */
+		#pragma omp parallel for if (get_global_parallel()->get_num_threads()!=1)
 		for (index_t i=0; i <num_subsets; ++i)
 		{
+			CMachine* machine;
+			CFeatures* features;
+			CLabels* labels;
+
+			if (get_global_parallel()->get_num_threads()==1)
+			{
+				machine=m_machine;
+			}
+			else
+			{
+				machine=(CMachine*)m_machine->clone();
+				machine->set_store_model_features(true);
+			}
 			/* evtl. update xvalidation output class */
 			CCrossValidationOutput* current=(CCrossValidationOutput*)
 					m_xval_outputs->get_first_element();
+			#pragma omp critical
+			{
 			while (current)
 			{
 				current->update_fold_index(i);
 				SG_UNREF(current);
 				current=(CCrossValidationOutput*)
 						m_xval_outputs->get_next_element();
 			}
+			}
 
 			/* set feature subset for training */
 			SGVector<index_t> inverse_subset_indices=
 					m_splitting_strategy->generate_subset_inverse(i);
-			m_features->add_subset(inverse_subset_indices);
-			for (index_t p=0; p<m_features->get_num_preprocessors(); p++)
+			
+			if (get_global_parallel()->get_num_threads()==1)
 			{
-				CPreprocessor* preprocessor = m_features->get_preprocessor(p);
-				preprocessor->init(m_features);
-				SG_UNREF(preprocessor);
+				features=m_features;
 			}
+			else
+				features=(CFeatures*)m_features->clone();
+			
+			features->add_subset(inverse_subset_indices);
 
 			/* set label subset for training */
-			m_labels->add_subset(inverse_subset_indices);
+			if (get_global_parallel()->get_num_threads()==1)
+			{
+				labels=m_labels;
+			}
+			else
+				labels=machine->get_labels();
```

Comment:
wrong tab indent. This automatically REFs, no need to decrease refcounter at the end